credentials/stscreds/assume_role_provider.go

```diff
@@ -0,0 +1,314 @@
+	creds := stscreds.NewAssumeRoleProvider(stscreds.AssumeRoleOptions{
+		Client: sts.NewFromConfig(cfg),
+		RoleARN: "myRoleArn",
+		SerialNumber: aws.String("myTokenSerialNumber"),
+		TokenCode: aws.String("00000000"),
+	})
+
+	cfg.Credentials = aws.CredentialsCache{Provider: creds}
+
+	// Create service client value configured for credentials
+	// from assumed role.
+	svc := s3.NewFromConfig(cfg)
+
+Assume Role with MFA Token Provider
+
+To assume an IAM role with MFA for longer running tasks where the credentials
+may need to be refreshed setting the TokenProvider field of AssumeRoleProvider
+will allow the credential provider to prompt for new MFA token code when the
+role's credentials need to be refreshed.
+
+The StdinTokenProvider function is available to prompt on stdin to retrieve
+the MFA token code from the user. You can also implement custom prompts by
+satisfying the TokenProvider function signature.
+
+Using StdinTokenProvider with multiple AssumeRoleProviders, or Credentials will
+have undesirable results as the StdinTokenProvider will not be synchronized. A
+single Credentials with an AssumeRoleProvider can be shared safely.
+
+	cfg, err := config.LoadDefaultConfig()
+	if err != nil {
+		panic(err)
+	}
+
+	// Create the credentials from AssumeRoleProvider to assume the role
+	// referenced by the "myRoleARN" ARN using the MFA token code provided.
+	creds := stscreds.NewAssumeRoleProvider(stscreds.AssumeRoleOptions{
+		Client: sts.NewFromConfig(cfg),
+		RoleARN: "myRoleArn",
+		SerialNumber: aws.String("myTokenSerialNumber"),
+		TokenProvider: stscreds.StdinTokenProvider,
+	})
+
+	cfg.Credentials = aws.CredentialsCache{Provider: creds}
+
+	// Create service client value configured for credentials
+	// from assumed role.
+	svc := s3.NewFromConfig(cfg)
+*/
+package stscreds
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/aws/aws-sdk-go-v2/service/sts/types"
+)
+
+// StdinTokenProvider will prompt on stdout and read from stdin for a string value.
+// An error is returned if reading from stdin fails.
+//
+// Use this function go read MFA tokens from stdin. The function makes no attempt
+// to make atomic prompts from stdin across multiple gorouties.
+//
+// Using StdinTokenProvider with multiple AssumeRoleProviders, or Credentials will
+// have undesirable results as the StdinTokenProvider will not be synchronized. A
+// single Credentials with an AssumeRoleProvider can be shared safely
+//
+// Will wait forever until something is provided on the stdin.
+func StdinTokenProvider() (string, error) {
+	var v string
+	fmt.Printf("Assume Role MFA token code: ")
+	_, err := fmt.Scanln(&v)
+
+	return v, err
+}
+
+// ProviderName provides a name of AssumeRole provider
+const ProviderName = "AssumeRoleProvider"
+
+// AssumeRole represents the minimal subset of the STS client API used by this provider.
+type AssumeRole interface {
```

Comment:
What do you think about `AssumeRoleAPIClient` for this? I think a name similar to that was proposed for higher level features like paginators would use.